File: pascal/is_prime.pas
program IsPrime;

uses
  sysutils,
  math;

type
  BoolArray = Array of Boolean;

function Sieve(N: LongInt): BoolArray;
var
  I, J: LongInt;
begin
  SetLength(Sieve, N + 1);
  for I := 2 to N do
    Sieve[I] := True;
  for I := 2 to N do
    begin
      if not Sieve[I] then
        continue;
      J := I * I;
      while J <= N do
        begin
          Sieve[J] := False;
          J := J + I;
        end;
    end;
end;

var
  N, I, Max: LongInt;
  Values: Array of LongInt;
  PrimeTable: BoolArray;
begin
  N := ParamCount();
  if N = 0 then
    exit;

  SetLength(Values, N);
  for I := 0 to N - 1 do
    Values[I] := StrToInt(ParamStr(I + 1));

  Max := MaxValue(Values);
  PrimeTable := Sieve(Max);

  for I in Values do
    begin
      if PrimeTable[I] then
        Write('Yes ')
      else
        Write('No ');
    end;
  WriteLn('');
end.

File: rust/src/bin/is_prime.rs
use std::env;

fn sieve(n: usize) -> Vec<bool> {
    let mut v = vec![true; n + 1];
    v[0] = false;
    v[1] = false;
    for i in 2 .. v.len() {
        if v[i] == false {
            continue;
        }
        let mut j = i * i;
        while j < v.len() {
            v[j] = false;
            j += i;
        }
    }
    return v;
}

fn main() {
    let values: Vec<usize> = env::args().skip(1).map(|s: String| s.parse().unwrap()).collect();
    if values.len() == 0 {
        return;
    }
    let max = values.iter().max().unwrap().clone();
    let prime_table = sieve(max);
    for i in &values {
        if prime_table[*i] {
            print!("Yes ");
        } else {
            print!("No ");
        }
    }
    println!("");
}

